middle_end/flambda2/from_lambda/lambda_to_lambda_transforms.ml

```diff
@@ -33,6 +33,53 @@ let switch_for_if_then_else ~cond ~ifso ~ifnot ~kind =
   in
   L.Lswitch (cond, switch, L.try_to_find_location cond, kind)
 
+(* This function helps bind expression to avoid duplicating them in the
+   generated code. Currently, this is only used for if-then-else optimization,
+   which guarantees that even if expressions are duplicated, they are still only
+   evaluated once (but could have been duplicated among multiple code paths), so
+   this is for code size optimization mainly.
+
+   One could want to duplicate constants, however the backend does not
+   deduplicate them, especially when they are the return value of the function,
+   so for now constants returned are shared via a static catch/raise. *)
+let bind_expr ~kind ~expr k =
+  let is_simple_duplicable expr =
+    match[@warning "-4"] (expr : L.lambda) with
+    | Lvar _ | Lconst _ -> true
+    | _ -> false
+  in
+  match[@warning "-4"] (expr : L.lambda) with
+  | Lstaticraise (_, args) when List.for_all is_simple_duplicable args -> k expr
+  | _ ->
+    let cont = L.next_raise_count () in
+    let jump = L.Lstaticraise (cont, []) in
+    L.Lstaticcatch (k jump, (cont, []), expr, Same_region, kind)
+
+let optimize_boolean_sequop ~cond ~ifso ~ifnot ~kind =
+  let rec aux ~kind ~cond ~ifso ~ifnot =
+    match[@warning "-4"] cond with
+    (* CR gbury: should we try to use the locs here, or is it better to keep
+       using the locs from each individual condition ? *)
+    | L.Lprim (Psequand, [a; b], _loc) ->
+      bind_expr ~kind ~expr:(aux ~kind ~cond:b ~ifso ~ifnot) (fun ifso ->
+          aux ~kind ~cond:a ~ifso ~ifnot)
+    | L.Lprim (Psequor, [a; b], _loc) ->
+      bind_expr ~kind ~expr:(aux ~kind ~cond:b ~ifso ~ifnot) (fun ifnot ->
+          aux ~kind ~cond:a ~ifso ~ifnot)
+    | L.Lprim (Pnot, [c], _loc) -> aux ~kind ~cond:c ~ifso:ifnot ~ifnot:ifso
+    | L.Lprim ((Psequand | Psequor), _, _) ->
+      Misc.fatal_error "Psequand / Psequor must have exactly two arguments"
```

Comment:
It's a bit weird to see this error here, in particular since I believe that it is too late: the code in `transform_primitive0` already matches on lists with two elements, so if we end up with a wrong number of elements this case will not even be reached (unless it is deep in a sequence). Instead, we will generate an invalid `Primitive` which will get caught later (possibly with a different error message).